src/controller/CHIPDeviceController.cpp

```diff
@@ -945,6 +961,53 @@ void DeviceCommissioner::OnSessionEstablished()
     RendezvousCleanup(CHIP_NO_ERROR);
 }
 
+CHIP_ERROR DeviceCommissioner::SendOperationalCertificateSigningRequestCommand(NodeId remoteDeviceId)
+{
+    // TODO: Call OperationalCredentials cluster API to send command
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceCommissioner::OnOperationalCertificateSigningRequest(NodeId node, const uint8_t * csrBuf, uint32_t csrLength)
+{
+    ReturnErrorCodeIf(mOperationalCredentialsDelegate == nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    uint8_t opCert[kMaxCHIPOpCertLength];
+    uint32_t opCertLen = 0;
+    ReturnErrorOnFailure(mOperationalCredentialsDelegate->GenerateNodeOperationalCertificate(node, 0, csrBuf, csrLength, 0, opCert,
+                                                                                             kMaxCHIPOpCertLength, opCertLen));
+
+    uint8_t signingCert[kMaxCHIPOpCertLength];
+    uint32_t signingCertLen = 0;
+    CHIP_ERROR err =
+        mOperationalCredentialsDelegate->GetIntermediateCACertificate(0, signingCert, kMaxCHIPOpCertLength, signingCertLen);
+    if (err == CHIP_ERROR_NOT_IMPLEMENTED)
+    {
+        err            = CHIP_NO_ERROR;
+        signingCertLen = 0;
+    }
+    ReturnErrorOnFailure(err);
+
+    ReturnErrorOnFailure(SendOperationalCertificate(node, opCert, opCertLen, signingCert, signingCertLen));
+
+    ReturnErrorOnFailure(
+        mOperationalCredentialsDelegate->GetRootCACertificate(0, signingCert, kMaxCHIPOpCertLength, signingCertLen));
+
+    return SendTrustedRootCertificate(node, signingCert, signingCertLen);
+}
+
+CHIP_ERROR DeviceCommissioner::SendOperationalCertificate(NodeId remoteDeviceId, const uint8_t * opCertBuf, uint32_t opCertLength,
```

Comment:
I am grateful to see this functionality being introduced.  So please do not let my comment detract from that.

But I'm curious why the commissioner interfaces assume the commissioner is only on one fabric.  In the spec, this is all very symmetric: any given node can be on any number of fabrics.  When the other node receives this request, it cannot make an assumption about fabrics; it will route through a session context to figure out which admin domain this transaction belongs to.

I presume that the answer is that a constructed commissioner object _does_ belong to a single fabric.  Does this mean though that the commissioner constructor will be passed some sort of fabric identifier or cert?  If not, how will the commissioner know which fabric it is associated with?  Will the SDK support construction of multiple commissioner objects?

It would seem an odd and unfortunate constraint if the presumably much more capable commissioner nodes end up constrained to a single fabric, whereas typically-constrained accessory devices must support multiple fabrics.  Again, in the spec this is symmetric.

Reply:
@msandstedt , the current design is geared towards what you mentioned in this statement: `I presume that the answer is that a constructed commissioner object does belong to a single fabric`.

The fabric provisioning is not yet complete. So, we can tweak the APIs to fit the requirements. Do we foresee one commissioner application catering to multiple fabrics?

Reply:
I should disambiguate between commissioner and controller.

I think a single controller device on multiple fabrics is an important and useful construct.  Multiple controllers could be bound to separate ports, but the sdk isn't currently well-structured to support that.  But it's also not necessary as the spec provides for all demuxing to happen through session context.  

I would like the same to be true of commissioners.  However, user-directed-commissioning precludes this without each commissioner instance listening to a unique port.  This is due to user-directed-commissioning, which requires the commissioner to provide a service at its listening port to receive `IdentificationDeclaration` messages that are implicitly directed to a single fabric.  I imagine it is plausible that other services will be added over time.  Again, the sdk isn't well-structured to allow a single CHIP process to listen to multiple ports.  Multiple concurrent commissioners therefore appears like a heavier lift than multiple concurrent controllers.  But the spec in no way precludes this.

What I would really like to avoid is the need for integrators to segment multiple controller or commissioner instances into separate processes.  Nothing in the spec gives rise to such a requirement.  If we do end up in that situation, it would only be through design decisions that maybe did not take into account these use cases.  That is why I'm bringing it up.

Reply:
@msandstedt this are good observations. Maybe you could submit GitHub issues for each of the requirements. That way the discussion won't be lost once this PR merges. And, it'll help prioritize the design/SDK updates.

Reply:
Sure, will do.

Reply:
Filed here:

https://github.com/project-chip/connectedhomeip/issues/6541
https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/2918